parser-typechecker/src/Unison/Util/EnumContainers.hs

```diff
@@ -169,3 +177,36 @@ traverseWithKey f (EM m) = EM <$> IM.traverseWithKey (f . intToKey) m
 
 setSize :: EnumSet k -> Int
 setSize (ES s) = IS.size s
+
+-- | A type optimized for lookups by enum for small numbers of key-value pairs.
+--
+-- It aims to minimize pointer-chasing and pack memory tightly so it fits in cache-lines for the CPU
+data SmallEnumMap k v
+  = SmallEnumMap
+      {-# UNPACK #-} !(VUnboxed.Vector Int) -- keys
+      {-# UNPACK #-} !(Vector.Vector v) -- values
+  deriving (Show, Eq, Ord, Functor, Foldable, Traversable)
+
+mapToSmallEnumMap :: (EnumKey k) => EnumMap k v -> SmallEnumMap k v
+mapToSmallEnumMap !(EM m) =
+  let (keys, values) = unzip $ IM.toList m
+   in SmallEnumMap (VUnboxed.fromList keys) (Vector.fromList values)
+
+smallEnumMapFromList :: (EnumKey k) => [(k, v)] -> SmallEnumMap k v
+smallEnumMapFromList kvs =
+  let (keys, values) = unzip kvs
+   in SmallEnumMap (VUnboxed.fromList $ fmap keyToInt keys) (Vector.fromList values)
+
+smallEnumMapElems :: SmallEnumMap k v -> Vector.Vector v
+smallEnumMapElems !(SmallEnumMap _ v) = v
+
+smallEnumMapToList :: (EnumKey k) => SmallEnumMap k v -> [(k, v)]
+smallEnumMapToList !(SmallEnumMap keys values) =
+  zip (fmap intToKey . VUnboxed.toList $ keys) (Vector.toList values)
+
+smallEnumMapLookup :: (EnumKey k) => k -> SmallEnumMap k v -> Maybe v
+smallEnumMapLookup !k !(SmallEnumMap keys values) =
```

Comment:
Hmm. I was thinking we should optimize the case where the map has contiguous integer keys, in which case you do a single vector lookup to resolve the branch. No elemIndex happening. Did you try this?

Reply:
Yes that's the one I tried first with similarly poor results, I think the common pattern-matching case is just so low arity that it doesn't seem to make much difference.

I also tried optimizing the single branch and 2 branch cases.